akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -293,38 +293,108 @@ private[akka] object StreamLayout {
-    realPublisher = RejectAdditionalSubscribers[T]
-    sub
+    requireNonNullSubscriber(s)
+    if (subscriptionStatus.compareAndSet(null, s)) () // wait for onSubscribe
+    else
+      subscriptionStatus.get match {
+        case sub: Subscriber[_] ⇒ rejectAdditionalSubscriber(s, "VirtualProcessor")
+        case sub: Sub ⇒
+          try {
+            subscriptionStatus.set(s)
+            tryOnSubscribe(s, sub)
+            sub.closeLatch() // allow onNext only now
+            terminationStatus.getAndSet(Allowed) match {
+              case null       ⇒ // nothing happened yet
+              case Completed  ⇒ tryOnComplete(s)
+              case Failed(ex) ⇒ tryOnError(s, ex)
+              case Allowed    ⇒ // all good
+            }
+          } catch {
+            case NonFatal(ex) ⇒ sub.cancel()
+          }
+      }
+  }
+
+  override def onSubscribe(s: Subscription): Unit = {
+    requireNonNullSubscription(s)
+    val wrapped = new Sub(s)
+    if (subscriptionStatus.compareAndSet(null, wrapped)) () // wait for Subscriber
+    else
+      subscriptionStatus.get match {
+        case sub: Subscriber[_] ⇒
+          terminationStatus.get match {
+            case Allowed ⇒
+              /*
+               * There is a race condition here: if this thread reads the subscriptionStatus after
+               * set set() in subscribe() but then sees the terminationStatus before the getAndSet()
+               * is published then we will rely upon the downstream Subscriber for cancelling this
+               * Subscription. I only mention this because the TCK requires that we handle this here
+               * (since the manualSubscriber used there does not expose this behavior).
+               */
+              s.cancel()
+            case _ ⇒
+              tryOnSubscribe(sub, wrapped)
+              wrapped.closeLatch() // allow onNext only now
+              terminationStatus.set(Allowed)
+          }
+        case sub: Subscription ⇒
+          s.cancel() // reject further Subscriptions
+      }
+  }
+
+  override def onError(t: Throwable): Unit = {
+    requireNonNullException(t)
+    if (terminationStatus.compareAndSet(null, Failed(t))) () // let it be picked up by subscribe()
+    else tryOnError(subscriptionStatus.get.asInstanceOf[Subscriber[T]], t)
+  }
+
+  override def onComplete(): Unit =
+    if (terminationStatus.compareAndSet(null, Completed)) () // let it be picked up by subscribe()
+    else tryOnComplete(subscriptionStatus.get.asInstanceOf[Subscriber[T]])
+
+  override def onNext(t: T): Unit = {
+    requireNonNullElement(t)
+    tryOnNext(subscriptionStatus.get.asInstanceOf[Subscriber[T]], t)
+  }
+
+  private final class Sub(s: Subscription) extends AtomicLong with Subscription {
+    override def cancel(): Unit = {
+      subscriptionStatus.set(InertSubscriber)
+      s.cancel()
+    }
+    @tailrec
+    override def request(n: Long): Unit = {
```

Comment:
Without this the TCK fails: it rightfully signals reentrant signaling towards the Subscriber since requesting from onSubscribe may now immediately trigger the asynchronous invocation of onNext. Therefore requests during onSubscribe are delayed by this method.